### PR TITLE
Precomputed velocity

### PR DESCRIPTION
**Pull request by clusty**
### Purpose of this PR
Backport of https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/3981 , https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/3996
The option was renamed to **Add Precomputed Velocity** as discussed with @sebastienlagarde 

---
### Release Notes
Add option to have Additional PrecomputedVelocity (Useful for Alembic)

---
### Testing status
**Katana Tests**: First off we need to make sure the Katana SRP tests are green?
Katana is green: https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=precomputedVelocity&automation-tools_branch=master&unity_branch=2019.2%2Fstaging&sort=4-asc
Yamato has failing tests:

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
Playmode tests on HDRP_Tests fail on my machine for a ton of strange reasons (no base image to compare with, etc). The release/2019.2 branch fails itself on my setup.
- [x] Built a player
- [x] Checked new UI names with UX convention
- [x] Tested UI multi-edition + Undo/Redo
- [x] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in devloper mode, you have a button at end of resources that check the pathes)
- Other: 
-Built shader Graph material based on lit, checked motion vectors work. Used regular a few regular materials and they look the same as before, and motion vectors work
-Built standalone OSX app

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

Launch Katana (Link on master here - update for your branch):
https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=master&automation-tools_branch=master&unity_branch=trunk&sort=1-asc

Alternative, launch Yamato (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline

Any test projects to go with this to help reviewers?

---
### Overall Product Risks
**Technical Risk**: Medium
New feature, disabled by default

**Halo Effect**:Low

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Editor/Material/Fabric/ShaderGraph/FabricSettingsView.cs [hunk 1]
```diff
@@ -365,6 +374,14 @@ void ChangeSSR(ChangeEvent<bool> evt)
             m_Node.receiveSSR = td;
         }
 
+        void ChangeAddPrecomputedVelocity(ChangeEvent<bool> evt)
+        {
+            m_Node.owner.owner.RegisterCompleteObjectUndo("Add Velocity Change");
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Editor/Material/Fabric/ShaderGraph/FabricSettingsView.cs]: Please change this to Add Precomputed Velocity

File: com.unity.render-pipelines.high-definition/Editor/Material/Fabric/ShaderGraph/FabricSubShader.cs [hunk 2]
```diff
@@ -286,8 +286,7 @@ class FabricSubShader : IFabricSubShader
                 HDSubShaderUtilities.SetStencilStateForForward(ref pass);
                 HDSubShaderUtilities.SetBlendModeForForward(ref pass);
 
-                pass.ExtraDefines.Remove("#ifndef DEBUG_DISPLAY\n#define SHADERPASS_FORWARD_BYPASS_ALPHA_TEST\n#endif");                
-
+                pass.ExtraDefines.Remove("#ifndef DEBUG_DISPLAY\n#define SHADERPASS_FORWARD_BYPASS_ALPHA_TEST\n#endif");
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Editor/Material/Fabric/ShaderGraph/FabricSubShader.cs]: Please remove this white space change.

File: com.unity.render-pipelines.high-definition/Editor/Material/Lit/ShaderGraph/HDLitSettingsView.cs [hunk 3]
```diff
@@ -586,6 +595,14 @@ void ChangeSSR(ChangeEvent<bool> evt)
             td.isOn = evt.newValue;
             m_Node.receiveSSR = td;
         }
+        
+        void ChangeAddPrecomputedVelocity(ChangeEvent<bool> evt)
+        {
+            m_Node.owner.owner.RegisterCompleteObjectUndo("Add Velocity Change");
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Editor/Material/Lit/ShaderGraph/HDLitSettingsView.cs]: Here also please change to Add Precomputed Velocity

File: com.unity.render-pipelines.high-definition/Editor/Material/StackLit/ShaderGraph/StackLitSettingsView.cs [hunk 4]
```diff
@@ -632,6 +642,14 @@ void ChangeReceiveSSR(ChangeEvent<bool> evt)
             m_Node.receiveSSR = td;
         }
 
+        void ChangeAddPrecomputedVelocity(ChangeEvent<bool> evt)
+        {
+            m_Node.owner.owner.RegisterCompleteObjectUndo("Add Velocity Change");
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Editor/Material/StackLit/ShaderGraph/StackLitSettingsView.cs]: Same as above change to "Add Precomputed Velocity"

File: com.unity.render-pipelines.high-definition/Editor/Material/UIBlocks/ShaderGraphUIBlock.cs [hunk 5]
```diff
@@ -166,6 +168,19 @@ void DrawMotionVectorToggle()
             bool enabled = materials[0].GetShaderPassEnabled(HDShaderPassNames.s_MotionVectorsStr);
             EditorGUI.BeginChangeCheck();
             enabled = EditorGUILayout.Toggle("Motion Vector For Vertex Animation", enabled);
+            
+            // We would like to automatically enable the motion vector pass (handled on material UI side)
+            // in case we have additional velocity change enabled in a graph. Due to serialization of material, changing
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Editor/Material/UIBlocks/ShaderGraphUIBlock.cs]: Rename to precomputed velocity

File: com.unity.render-pipelines.high-definition/Runtime/Material/AxF/AxF.shader [hunk 6]
```diff
@@ -110,6 +110,7 @@ Shader "HDRP/AxF"
 
         [ToggleUI] _SupportDecals("Support Decals", Float) = 1.0
         [ToggleUI] _ReceivesSSR("Receives SSR", Float) = 1.0
+        [ToggleUI] _AddPrecomputedVelocity("EnableAdditionalVelocity", Float) = 0.0
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Runtime/Material/AxF/AxF.shader]: Please rename to AddPrecomputedVelocity

File: com.unity.render-pipelines.high-definition/Runtime/Material/LayeredLit/LayeredLit.shader [hunk 7]
```diff
@@ -364,6 +364,7 @@ Shader "HDRP/LayeredLit"
 
         [ToggleUI] _SupportDecals("Support Decals", Float) = 1.0
         [ToggleUI] _ReceivesSSR("Receives SSR", Float) = 1.0
+        [ToggleUI] _AddPrecomputedVelocity("EnableAdditionalVelocity", Float) = 0.0
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Runtime/Material/LayeredLit/LayeredLit.shader]: Please rename to AddPrecomputedVelocity

File: com.unity.render-pipelines.high-definition/Runtime/Material/LayeredLit/LayeredLitTessellation.shader [hunk 8]
```diff
@@ -374,6 +374,7 @@ Shader "HDRP/LayeredLitTessellation"
 
         [ToggleUI] _SupportDecals("Support Decals", Float) = 1.0
         [ToggleUI] _ReceivesSSR("Receives SSR", Float) = 1.0
+        [ToggleUI] _AddPrecomputedVelocity("EnableAdditionalVelocity", Float) = 0.0
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Runtime/Material/LayeredLit/LayeredLitTessellation.shader]: Please rename to AddPrecomputedVelocity

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/Lit.shader [hunk 9]
```diff
@@ -220,7 +220,8 @@ Shader "HDRP/Lit"
 
         [ToggleUI] _SupportDecals("Support Decals", Float) = 1.0
         [ToggleUI] _ReceivesSSR("Receives SSR", Float) = 1.0
-
+        [ToggleUI] _AddPrecomputedVelocity("EnableAdditionalVelocity", Float) = 0.0
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Lit/Lit.shader]: Please rename to AddPrecomputedVelocity

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitTessellation.shader [hunk 10]
```diff
@@ -231,6 +231,7 @@ Shader "HDRP/LitTessellation"
 
         [ToggleUI] _SupportDecals("Support Decals", Float) = 1.0
         [ToggleUI] _ReceivesSSR("Receives SSR", Float) = 1.0
+        [ToggleUI] _AddPrecomputedVelocity("EnableAdditionalVelocity", Float) = 0.0
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Lit/LitTessellation.shader]: Same as above please rename to AddPrecomputedVelocity

File: com.unity.render-pipelines.high-definition/Runtime/Material/TerrainLit/TerrainLit.shader [hunk 11]
```diff
@@ -45,6 +45,7 @@ Shader "HDRP/TerrainLit"
 
         [HideInInspector] [ToggleUI] _SupportDecals("Support Decals", Float) = 1.0
         [HideInInspector] [ToggleUI] _ReceivesSSR("Receives SSR", Float) = 1.0
+        [HideInInspector] [ToggleUI] _AddPrecomputedVelocity("EnableAdditionalVelocity", Float) = 0.0
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Runtime/Material/TerrainLit/TerrainLit.shader]: Same as above please rename to AddPrecomputedVelocity

File: com.unity.render-pipelines.high-definition/Runtime/Material/Unlit/Unlit.shader [hunk 12]
```diff
@@ -68,6 +68,9 @@ Shader "HDRP/Unlit"
         // Motion vector pass
         [HideInInspector] _StencilRefMV("_StencilRefMV", Int) = 128 // StencilBitMask.ObjectMotionVectors
         [HideInInspector] _StencilWriteMaskMV("_StencilWriteMaskMV", Int) = 128 // StencilBitMask.ObjectMotionVectors
+
+        [ToggleUI] _AddPrecomputedVelocity("EnableAdditionalVelocity", Float) = 0.0
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Unlit/Unlit.shader]: As above

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/MotionVectorVertexShaderCommon.hlsl [hunk 13]
```diff
@@ -121,16 +124,19 @@ PackedVaryingsType MotionVectorVS(inout VaryingsType varyingsType, AttributesMes
     else
     {
         bool hasDeformation = unity_MotionVectorsParams.x > 0.0; // Skin or morph target
+        float3 effectivePositionOS = (hasDeformation ? inputPass.previousPositionOS : inputMesh.positionOS);
+ #if defined(_ADD_PRECOMPUTED_VELOCITY)
+         effectivePositionOS -= inputPass.alembicVelocity; // <= this line
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/MotionVectorVertexShaderCommon.hlsl]: Not yours but we could remove the comment // <= this line

File: com.unity.render-pipelines.high-definition/Editor/Material/StackLit/ShaderGraph/StackLitSettingsView.cs [hunk 14]
```diff
@@ -644,7 +644,7 @@ void ChangeReceiveSSR(ChangeEvent<bool> evt)
 
         void ChangeAddPrecomputedVelocity(ChangeEvent<bool> evt)
         {
-            m_Node.owner.owner.RegisterCompleteObjectUndo("Add Velocity Change");
+            m_Node.owner.owner.RegisterCompleteObjectUndo("Add Precomputed Velocity");
```
Comment (jeanblouin) [inline com.unity.render-pipelines.high-definition/Editor/Material/StackLit/ShaderGraph/StackLitSettingsView.cs]: Thank you for the change

**Comment by sebastienlagarde** [comment]
Renaming for 2019.3 is in this PR: https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/4280

**Pull request merged**